jbmc/src/java_bytecode/java_object_factory.cpp

```diff
@@ -1010,37 +1039,62 @@ void java_object_factoryt::gen_nondet_struct_init(
         ::implements_java_char_sequence(struct_type);
     const bool has_length_and_data =
       struct_type.has_component("length") && struct_type.has_component("data");
+    const bool has_string_input_values =
+      !object_factory_parameters.string_input_values.empty();
 
-    // Add an initial all-zero write. Most of the fields of this will be
-    // overwritten, but it helps to have a whole-structure write that analysis
-    // passes can easily recognise leaves no uninitialised state behind.
-
-    // This code mirrors the `remove_java_new` pass:
-    auto initial_object = zero_initializer(struct_type, source_locationt(), ns);
-    CHECK_RETURN(initial_object.has_value());
-    irep_idt qualified_clsid = "java::" + id2string(class_identifier);
-    set_class_identifier(
-      to_struct_expr(*initial_object), ns, struct_tag_typet(qualified_clsid));
-
-    // If the initialised type is a special-cased String type (one with length
-    // and data fields introduced by string-library preprocessing), initialise
-    // those fields with nondet values
-    if(is_char_sequence && has_length_and_data)
-    {
+    if(is_char_sequence && has_length_and_data && has_string_input_values)
+    { // We're dealing with a string and we should set fixed input values.
       skip_special_string_fields = true;
 
-      initialize_nondet_string_fields(
-        to_struct_expr(*initial_object),
-        assignments,
-        object_factory_parameters.min_nondet_string_length,
-        object_factory_parameters.max_nondet_string_length,
+      // We create a switch statement where each case is an assignment
+      // of one of the fixed input strings to the input variable in question
+
+      alternate_casest cases =
+        get_string_input_values_code(
+          expr,
+          object_factory_parameters.string_input_values,
+          symbol_table);
+      assignments.add(generate_nondet_switch(
+        id2string(object_factory_parameters.function_id),
+        cases,
+        java_int_type(),
+        ID_java,
         loc,
-        object_factory_parameters.function_id,
-        symbol_table,
-        object_factory_parameters.string_printable);
+        symbol_table));
     }
+    else
+    {
+      // Add an initial all-zero write. Most of the fields of this will be
+      // overwritten, but it helps to have a whole-structure write that analysis
+      // passes can easily recognise leaves no uninitialised state behind.
```

Comment:
Me too :-) The sentence was there before, I've only put it in an else block and hence indentation changed. I don't know how to improve the sentence since I don't have good enough understanding of the code.